client/src/components/graph/graph.tsx

```diff
@@ -1033,68 +1054,98 @@ class Graph extends React.Component<GraphProps, GraphState> {
       const offscreenCanvas = document.createElement("canvas");
       offscreenCanvas.width = width;
       offscreenCanvas.height = height;
-      const legendNodes =
-        document.getElementById("continuous_legend")?.childNodes;
+
       const ctx = offscreenCanvas.getContext("2d");
       ctx?.drawImage(graph, 0, 0);
 
-      legendNodes?.forEach(async (node) => {
-        if (node.nodeName === "CANVAS") {
-          // get style from node
-          const style = window.getComputedStyle(node as HTMLCanvasElement);
-          const nodeWidth = parseInt(style.width, 10);
-          const nodeHeight = parseInt(style.height, 10);
-          const nodeLeft = parseInt(style.left, 10);
-          const nodeTop = parseInt(style.top, 10);
-          ctx?.drawImage(
-            node as HTMLCanvasElement,
-            PADDING + nodeLeft,
-            PADDING + nodeTop,
-            nodeWidth,
-            nodeHeight
+      let categoricalLegendImageURI: string | null = null;
+
+      if (colors.colorMode !== null) {
+        if (colors.colorMode !== "color by categorical metadata") {
+          const legendNodes =
```

Comment:
should we extract these functions out of `graph.tsx` and import them here to use? I think it'll help us tame the readability of this giant file!